File: kaggle/ml_qa/ece_honesty.py
"""ECE residual honesty (M2) — advisory calibration residual for product UI.

Reads published metrics (e.g. E20 ``test_ece``) and classifies residual severity.
Never unlocks Identify. Never forage / consumption permission.

Bands (orientation-only heuristics for mycological multi-class):
  - good:     ECE < 0.05  — confidence roughly trustworthy for ranking chrome
  - moderate: ECE < 0.12  — show confidence only with de-emphasis
  - high:     ECE ≥ 0.12  — hide or heavily de-emphasize confidence on Identify
  - unknown:  missing metric

E20 source-holdout published ECE ≈ 0.188 → **high** residual (soft gates on MAP/deadly
can still PASS while confidence chrome must stay humble).
"""
from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

POLICY = "orientation_only_never_consume"
PRODUCT_UNLOCK = False

# Advisory bands — never product gates for unlock
ECE_GOOD_MAX = 0.05
ECE_MODERATE_MAX = 0.12

ROOT = Path(__file__).resolve().parents[2]
DEFAULT_METRICS = ROOT / "kaggle" / "kernel_output_v20" / "models" / "metrics.json"
DEFAULT_OUT = ROOT / "eval" / "reports" / "ml_experiments" / "e20_ece_residual.json"


def classify_ece_band(ece: float | None) -> str:
    if ece is None:
        return "unknown"
    try:
        v = float(ece)
    except (TypeError, ValueError):
        return "unknown"
    if v != v:  # NaN
        return "unknown"
    if v < 0:
        return "unknown"
    if v < ECE_GOOD_MAX:
        return "good"
    if v < ECE_MODERATE_MAX:
        return "moderate"
    return "high"


def ece_product_guidance(band: str) -> dict[str, Any]:
    """What Identify/product should do — never edible clearance."""
    if band == "good":
        return {
            "show_confidence": True,
            "deemphasize_confidence": False,
            "prefer_open_set_abstain": True,
            "summary_es": "ECE bajo: confianza usable con cautela · sigue sin permiso de consumo.",
            "summary_en": "Low ECE: confidence usable with caution · still never consumption permission.",
        }
    if band == "moderate":
        return {
            "show_confidence": True,
            "deemphasize_confidence": True,
            "prefer_open_set_abstain": True,
            "summary_es": "ECE moderado: de-enfatizar % en UI · open-set y multi-vista mandan.",
            "summary_en": "Moderate ECE: de-emphasize % in UI · open-set and multi-view lead.",
        }
    if band == "high":
        return {
            "show_confidence": False,  # product may still gate via shouldShowConfidence
            "deemphasize_confidence": True,
            "prefer_open_set_abstain": True,
            "summary_es": (
                "ECE alto (residual): no confiar en % del modelo. "
                "Abstención open-set + multi-vista + revisión humana. "
                "Nunca permiso de consumo."
            ),
            "summary_en": (
                "High ECE residual: do not trust model %. "
                "Open-set abstain + multi-view + human review. "
                "Never consumption permission."
            ),
        }
    return {
        "show_confidence": False,
        "deemphasize_confidence": True,
        "prefer_open_set_abstain": True,
        "summary_es": "ECE desconocido: fall-closed — no mostrar confianza como certeza.",
        "summary_en": "Unknown ECE: fail-closed — do not show confidence as certainty.",
    }


def residual_actions(band: str, ece: float | None, temperature: float | None) -> list[str]:
    actions = [
        "Keep product_unlock=false until operator cycle (orientation only).",
        "Never treat confidence as edible clearance or forage permission.",
    ]
    if band == "high":
        actions.extend(
            [
                "De-emphasize or hide Identify confidence chrome (shouldShowConfidence soft path).",
                "Prefer open-set reject / needs_review over forced top-1 display.",
                "Optional: re-temperature on holdout logits if test_predictions.npz available.",
                "Grow multi-view same-specimen holdout before trusting confidence UI.",
            ]
        )
    elif band == "moderate":
        actions.extend(
            [
                "Show confidence only with orientation disclaimer.",
                "Monitor ECE after next training loop; do not relax open-set thr for chrome.",
            ]
        )
    elif band == "good":
        actions.append("Still keep deadly@3 + open-set gates ahead of confidence chrome.")
    else:
        actions.append("Publish test_ece on next eval artifact for residual tracking.")
    if temperature is not None:
        try:
            t = float(temperature)
            if t > 1.2:
                actions.append(
                    f"Temperature already scaled (T={t:.3f}); residual ECE may remain high on domain shift."
                )
        except (TypeError, ValueError):
            pass
    if ece is not None:
        try:
            actions.append(f"Recorded ECE={float(ece):.4f} band={band}.")
        except (TypeError, ValueError):
            pass
    return actions


def build_ece_residual_from_metrics(
    metrics: dict[str, Any] | None,
    *,
    source: str = "metrics.json",
) -> dict[str, Any]:
    metrics = metrics or {}
    ece_raw = metrics.get("test_ece")
    if ece_raw is None:
        ece_raw = metrics.get("ece")
    try:
        ece = float(ece_raw) if ece_raw is not None else None
    except (TypeError, ValueError):
        ece = None
    band = classify_ece_band(ece)
    guidance = ece_product_guidance(band)
    temp = metrics.get("temperature")
    try:
        temperature = float(temp) if temp is not None else None
    except (TypeError, ValueError):
        temperature = None

    map3 = metrics.get("test_map_at_3")
    deadly3 = metrics.get("safety_recall_deadly_at_3") or metrics.get("safety_recall_deadly")

    return {
        "product_unlock": PRODUCT_UNLOCK,
        "can_auto_unlock": False,
        "forage_permission": False,
        "consumption_permission": False,
        "policy": POLICY,
        "source": source,
        "test_ece": ece,
        "band": band,
        "bands": {
            "good_max": ECE_GOOD_MAX,
            "moderate_max": ECE_MODERATE_MAX,
            "note": "Advisory only — not unlock gates",
        },
        "temperature": temperature,
        "test_map_at_3": map3,
        "safety_recall_deadly_at_3": deadly3,
        "eval_protocol": metrics.get("eval_protocol"),
        "test_domain": metrics.get("test_domain"),
        "guidance": guidance,
        "residual_actions": residual_actions(band, ece, temperature),
        "honesty": (
            "Soft MAP/deadly gates can PASS while ECE remains high (domain shift). "
            "Confidence UI must stay humble. Orientation only."
        ),
        "generated": datetime.now(timezone.utc).isoformat(),
    }


def load_metrics_json(path: Path | None = None) -> tuple[dict[str, Any] | None, Path]:
    p = Path(path) if path else DEFAULT_METRICS
    if not p.is_file():
        return None, p
    try:
        data = json.loads(p.read_text(encoding="utf-8"))
        return data if isinstance(data, dict) else None, p
    except (OSError, json.JSONDecodeError):
        return None, p


def build_ece_residual_report(metrics_path: Path | None = None) -> dict[str, Any]:
    metrics, path = load_metrics_json(metrics_path)
    if metrics is None:
        out = build_ece_residual_from_metrics(None, source=str(path))
        out["status"] = "missing_metrics"
        out["metrics_path"] = str(path)
        out["exists"] = path.is_file()
        return out
    out = build_ece_residual_from_metrics(metrics, source=str(path))
    out["status"] = "ok"
    out["metrics_path"] = str(path)
    out["exists"] = True
    out["version"] = metrics.get("version")
    return out


def write_ece_residual_report(
    repo: Path | None = None,
    *,
    metrics_path: Path | None = None,
    out_dir: Path | None = None,
) -> dict[str, Any]:
    repo = Path(repo or ROOT)
    report = build_ece_residual_report(metrics_path)
    dest_dir = Path(out_dir) if out_dir else repo / "eval" / "reports" / "ml_experiments"
    dest_dir.mkdir(parents=True, exist_ok=True)
    dest = dest_dir / "e20_ece_residual.json"
    dest.write_text(json.dumps(report, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
    report["report_path"] = str(dest)
    return report


if __name__ == "__main__":
    r = write_ece_residual_report()
    print(json.dumps({k: r[k] for k in ("status", "test_ece", "band", "product_unlock", "report_path") if k in r}, indent=2))

File: frontend/src/lib/eceHonesty.ts
/**
 * ECE residual honesty (M2/M2.1) — pure FE mirror of kaggle/ml_qa/ece_honesty.py.
 * Advisory only · never product_unlock · never forage/consumption.
 *
 * Soft MAP/deadly gates can PASS while ECE stays high (E20 ≈0.188).
 * Identify confidence chrome must follow ECE residual, not only metrics_acceptable.
 */

export type EceBand = 'good' | 'moderate' | 'high' | 'unknown'

export const ECE_GOOD_MAX = 0.05
export const ECE_MODERATE_MAX = 0.12

/**
 * Published E20 source-holdout residual (metrics.json test_ece).
 * Used when live `/models/status.ece_residual` is unavailable.
 * product_unlock always false.
 */
export const E20_ECE_SNAPSHOT = {
  test_ece: 0.18781378849301697,
  band: 'high' as EceBand,
  product_unlock: false as const,
  protocol: 'source_holdout_e20',
}

export function classifyEceBand(ece: number | null | undefined): EceBand {
  if (ece == null || !Number.isFinite(ece) || ece < 0) return 'unknown'
  if (ece < ECE_GOOD_MAX) return 'good'
  if (ece < ECE_MODERATE_MAX) return 'moderate'
  return 'high'
}

export type EceGuidance = {
  showConfidence: boolean
  deemphasizeConfidence: boolean
  summaryEs: string
  summaryEn: string
}

export function eceProductGuidance(band: EceBand): EceGuidance {
  switch (band) {
    case 'good':
      return {
        showConfidence: true,
        deemphasizeConfidence: false,
        summaryEs:
          'ECE bajo: confianza usable con cautela · sigue sin permiso de consumo.',
        summaryEn:
          'Low ECE: confidence usable with caution · still never consumption permission.',
      }
    case 'moderate':
      return {
        showConfidence: true,
        deemphasizeConfidence: true,
        summaryEs:
          'ECE moderado: de-enfatizar % en UI · open-set y multi-vista mandan.',
        summaryEn:
          'Moderate ECE: de-emphasize % in UI · open-set and multi-view lead.',
      }
    case 'high':
      return {
        showConfidence: false,
        deemphasizeConfidence: true,
        summaryEs:
          'ECE alto (residual): no confiar en % del modelo. Abstención open-set + multi-vista + revisión humana. Nunca permiso de consumo.',
        summaryEn:
          'High ECE residual: do not trust model %. Open-set abstain + multi-view + human review. Never consumption permission.',
      }
    default:
      return {
        showConfidence: false,
        deemphasizeConfidence: true,
        summaryEs:
          'ECE desconocido: fail-closed — no mostrar confianza como certeza.',
        summaryEn:
          'Unknown ECE: fail-closed — do not show confidence as certainty.',
      }
  }
}

/**
 * Combine dual-signal gate (shouldShowConfidence) with ECE residual.
 * ECE can only tighten chrome further — never unlocks confidence when gate hides it.
 * Never product_unlock / never edible clearance.
 */
export function resolveIdentifyConfidenceChrome(
  gateAllowsConfidence: boolean,
  eceBand: EceBand = E20_ECE_SNAPSHOT.band,
): {
  show: boolean
  deemphasize: boolean
  hideNumericPercent: boolean
  band: EceBand
  noteEs: string
  noteEn: string
  productUnlock: false
} {
  const g = eceProductGuidance(eceBand)
  // Gate + ECE both required for full bars; high/unknown force hide
  const show = Boolean(gateAllowsConfidence && g.showConfidence)
  const deemphasize = !show || g.deemphasizeConfidence
  const hideNumericPercent =
    !show || eceBand === 'high' || eceBand === 'unknown' || deemphasize
  return {
    show,
    deemphasize,
    hideNumericPercent,
    band: eceBand,
    noteEs: g.summaryEs,
    noteEn: g.summaryEn,
    productUnlock: false,
  }
}

/** Short sticky line for ResultCard / Identify (locale-aware). */
export function eceConfidenceStickyLine(
  band: EceBand = E20_ECE_SNAPSHOT.band,
  locale?: string,
): string {
  const en = (locale || 'es').toLowerCase().startsWith('en')
  const g = eceProductGuidance(band)
  return en ? g.summaryEn : g.summaryEs
}

/** Normalize API ece_residual block for dashboard. */
export function normalizeEceResidual(raw: unknown): {
  band: EceBand
  ece: number | null
  summary: string
  productUnlock: false
  status: string
  source: 'live' | 'snapshot' | 'unknown'
} {
  const o = (raw && typeof raw === 'object' ? raw : {}) as Record<string, unknown>
  const ece =
    typeof o.test_ece === 'number'
      ? o.test_ece
      : typeof o.ece === 'number'
        ? o.ece
        : null
  const bandFromApi = String(o.band || '')
  const hasLive =
    bandFromApi === 'good' ||
    bandFromApi === 'moderate' ||
    bandFromApi === 'high' ||
    bandFromApi === 'unknown' ||
    ece != null
  const band: EceBand =
    bandFromApi === 'good' ||
    bandFromApi === 'moderate' ||
    bandFromApi === 'high' ||
    bandFromApi === 'unknown'
      ? bandFromApi
      : classifyEceBand(ece)
  const g = eceProductGuidance(band)
  const guidance = o.guidance as Record<string, unknown> | undefined
  const summary =
    (typeof guidance?.summary_es === 'string' && guidance.summary_es) ||
    (typeof guidance?.summary_en === 'string' && guidance.summary_en) ||
    g.summaryEs
  return {
    band,
    ece,
    summary,
    productUnlock: false,
    status: String(o.status || 'unknown'),
    source: hasLive ? 'live' : 'unknown',
  }
}

/**
 * Extract ECE band from full `/models/status` JSON (or partial).
 * Fail-soft: returns E20 snapshot band when missing — never unlocks.
 */
export function eceBandFromModelsStatus(status: unknown): {
  band: EceBand
  ece: number | null
  source: 'live' | 'snapshot'
  productUnlock: false
} {
  const root = (status && typeof status === 'object' ? status : {}) as Record<
    string,
    unknown
  >
  const residual = root.ece_residual
  const summary = (root.summary && typeof root.summary === 'object'
    ? root.summary
    : {}) as Record<string, unknown>

  if (residual && typeof residual === 'object') {
    const n = normalizeEceResidual(residual)
    if (n.status !== 'unavailable' && n.band !== 'unknown') {
      return {
        band: n.band,
        ece: n.ece,
        source: 'live',
        productUnlock: false,
      }
    }
    // unknown band but has numeric ece
    if (n.ece != null) {
      return {
        band: classifyEceBand(n.ece),
        ece: n.ece,
        source: 'live',
        productUnlock: false,
      }
    }
  }

  const sumEce =
    typeof summary.ece === 'number'
      ? summary.ece
      : typeof summary.test_ece === 'number'
        ? summary.test_ece
        : null
  const sumBand = String(summary.ece_band || '')
  if (
    sumBand === 'good' ||
    sumBand === 'moderate' ||
    sumBand === 'high' ||
    sumBand === 'unknown'
  ) {
    return {
      band: sumBand,
      ece: sumEce,
      source: 'live',
      productUnlock: false,
    }
  }
  if (sumEce != null) {
    return {
      band: classifyEceBand(sumEce),
      ece: sumEce,
      source: 'live',
      productUnlock: false,
    }
  }

  return {
    band: E20_ECE_SNAPSHOT.band,
    ece: E20_ECE_SNAPSHOT.test_ece,
    source: 'snapshot',
    productUnlock: false,
  }
}

const API_BASE = import.meta.env.VITE_API_URL || '/api'
const API_KEY = import.meta.env.VITE_API_KEY || ''

/** In-memory cache for Identify session (avoid hammering /models/status). */
let _eceBandCache: {
  band: EceBand
  ece: number | null
  source: 'live' | 'snapshot'
  fetchedAt: number
} | null = null

export const ECE_BAND_CACHE_TTL_MS = 60_000

/**
 * Live fetch of ECE residual band for Identify confidence chrome.
 * Fail-soft → E20 snapshot (high). Never product_unlock.
 */
export async function fetchEceBandForIdentify(
  signal?: AbortSignal,
): Promise<{
  band: EceBand
  ece: number | null
  source: 'live' | 'snapshot'
  productUnlock: false
}> {
  const now = Date.now()
  if (
    _eceBandCache &&
    now - _eceBandCache.fetchedAt < ECE_BAND_CACHE_TTL_MS
  ) {
    return {
      band: _eceBandCache.band,
      ece: _eceBandCache.ece,
      source: _eceBandCache.source,
      productUnlock: false,
    }
  }

  try {
    const headers: Record<string, string> = { Accept: 'application/json' }
    if (API_KEY) headers['X-API-Key'] = API_KEY
    const res = await fetch(`${API_BASE.replace(/\/$/, '')}/models/status`, {
      method: 'GET',
      signal,
      headers,
    })
    if (!res.ok) {
      return {
        band: E20_ECE_SNAPSHOT.band,
        ece: E20_ECE_SNAPSHOT.test_ece,
        source: 'snapshot',
        productUnlock: false,
      }
    }
    const data = await res.json()
    const parsed = eceBandFromModelsStatus(data)
    _eceBandCache = {
      band: parsed.band,
      ece: parsed.ece,
      source: parsed.source,
      fetchedAt: now,
    }
    return {
      band: parsed.band,
      ece: parsed.ece,
      source: parsed.source,
      productUnlock: false,
    }
  } catch {
    return {
      band: E20_ECE_SNAPSHOT.band,
      ece: E20_ECE_SNAPSHOT.test_ece,
      source: 'snapshot',
      productUnlock: false,
    }
  }
}

/** Test helper: clear session cache. */
export function clearEceBandCache(): void {
  _eceBandCache = null
}
